examples/pytorch/pytorch_ask_and_tell_checkpoint.py

```diff
@@ -0,0 +1,227 @@
+                if batch_idx * BATCHSIZE >= N_VALID_EXAMPLES:
+                    break
+                data, target = data.view(data.size(0), -1).to(DEVICE), target.to(DEVICE)
+                output = model(data)
+                # Get the index of the max log-probability.
+                pred = output.argmax(dim=1, keepdim=True)
+                correct += pred.eq(target.view_as(pred)).sum().item()
+
+        accuracy = correct / min(len(valid_loader.dataset), N_VALID_EXAMPLES)
+
+        trial.report(accuracy, epoch)
+
+        # Save optimization status. We should save the objective value because the process may be
+        # killed between saving the last model and recording the objective value to the storage.
+        torch.save(
+            {
+                "epoch": epoch,
+                "model_state_dict": model.state_dict(),
+                "optimizer_state_dict": optimizer.state_dict(),
+                "accuracy": accuracy,
+            },
+            os.path.join(path, "model.pt"),
+        )
+
+        # Handle pruning based on the intermediate value.
+        if trial.should_prune():
+            raise optuna.exceptions.TrialPruned()
+
+    return accuracy
+
+
+def restart_from_checkpoint(study, trial):
+    # Enqueue trial with the same parameters as the stale trial to use saved information.
+
+    path = f"pytorch_checkpoint/{trial.number}/model.pt"
+    user_attrs = copy.deepcopy(trial.user_attrs)
+    if os.path.exists(path):
+        user_attrs["checkpoint_path"] = path
+
+    study.add_trial(
+        optuna.create_trial(
+            state=optuna.trial.TrialState.WAITING,
+            params=trial.params,
+            distributions=trial.distributions,
+            user_attrs=user_attrs,
+            system_attrs=trial.system_attrs,
+        )
+    )
+
+
+if __name__ == "__main__":
+    storage = optuna.storages.RDBStorage(
+        "sqlite:///example.db", heartbeat_interval=1, failed_trial_callback=restart_from_checkpoint
+    )
+    study = optuna.create_study(
+        storage=storage, study_name="pytorch_checkpoint", direction="maximize", load_if_exists=True
+    )
+
+    N_TRIALS = 10
+    TIMEOUT = 600
+    start_time = datetime.datetime.now()
+
+    for _ in range(N_TRIALS):
+        optuna.study.run_failed_trial_callback(study)
+        trial = study.ask()
+        thread, stop_event = optuna.study.create_heartbeat_thread(study, trial)
+
+        try:
+            value = objective(trial)
+            study.tell(trial, value)
+            print(
+                f"Trial {trial.number} finished with value: {value} and parameters: "
+                f"{trial.params}. Best is trial {study.best_trial.number} with "
+                f"value: {study.best_value}."
+            )
+        except optuna.TrialPruned:
+            print(f"Trial {trial.number} was pruned.")
+            study.tell(trial, None, optuna.trial.TrialState.PRUNED)
+
+        optuna.study.join_heartbeat_thread(study, thread, stop_event)
```

Comment:
I think that this call should be placed within a `finally` block to prevent thread leaks.

Reply:
Furthermore, if possible, I prefer using python `Context` to create and stop those threads rather than bare function calls (as manual management seems more error-prone).

The code below is a pseudo code to share my thought:
```python
class TrialHeartbeat:
    def __init__(self, trial):
        self._trial = trial

    def __enter__(self):
        optuna.study.run_failed_trial_callback(self._trial.study)  # I think that it's okay to call this function if it's just after the trial creation.
        self._thread, sefl._stop_event = optuna.study.create_heartbeat_thread(self._trial.study, self._trial)
        return self

    def __exit__(self, ...):
       optuna.study.join_heartbeat_thread(sefl._trial.study, self._thread, self._stop_event)


# Usage
trial = study.ask()
with TrialHeartbeat(trial):
    ...
    study.tell(trial, objective_value)
```

Reply:
> Furthermore, if possible, I prefer using python `Context` to create and stop those threads rather than bare function calls (as manual management seems more error-prone).

Thank you for your suggestion. That sounds reasonable to me.

Users may sometimes calls `ask` and `tell` in different processes, but the current heartbeat related functions are supposed to be called in a single process.  `Context` will prevent users to call the functions in different processes.

I'll update the PR based on python `Context`.